### PR TITLE
Release/3.10.5

### PR DESCRIPTION
**Pull request by thoma202**
See #29  + Contract and BNCodes

### CODE REVIEW

File: controllers/front/submitplus.php [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+                        );
+                        $return['success'][] = 'Votre paiement a bien été pris en compte.';
+                    } else {
+
+                        $paypal->validateOrder(
+                            $this->id_cart, $this->getOrderStatus('payment_error'),
+                            $payment->transactions[0]->amount->total, $payment->payer->payment_method, null,
+                            $transaction
+                        );
+                        $return['error'][] = 'Une erreur est survenue pendant le paiement.';
```
Comment (Quetzacoalt91) [inline controllers/front/submitplus.php]: Displayed text should be written first in English and then translated.

You have already everything you need for that with `$paypal->l(...)`

File: controllers/front/submitplus.php [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+                    }
+                } elseif ($submit == 'confirmCancel') {
+
+                    $paypal->validateOrder(
+                        $this->id_cart, $this->getOrderStatus('order_canceled'),
+                        $payment->transactions[0]->amount->total, $payment->payer->payment_method, null, $transaction
+                    );
+                    $return['success'][] = 'Votre commande a bien été annulé.';
```
Comment (Quetzacoalt91) [inline controllers/front/submitplus.php]: Text should be translated

File: controllers/front/submitplus.php [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+                } else
+                    $return['error'][] = 'Une erreur est survenue pendant le paiement.';
```
Comment (Quetzacoalt91) [inline controllers/front/submitplus.php]: Text should be translated

File: controllers/front/submitplus.php [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            } else
+                $return['error'][] = 'Une erreur est survenue pendant le paiement.';
```
Comment (Quetzacoalt91) [inline controllers/front/submitplus.php]: Text should be translated

File: controllers/front/submitplus.php [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        } else
+            $return['error'][] = 'Une erreur est survenue pendant le paiement.';
```
Comment (Quetzacoalt91) [inline controllers/front/submitplus.php]: Text should be translated

File: controllers/front/submitplus.php [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+    public function getOrderStatus($template)
+    {
+        /*
+         * payment
+         * payment_error
+         * order_canceled
+         * refund
+         */
+        return Db::getInstance()->getValue('SELECT id_order_state FROM '._DB_PREFIX_.'order_state_lang WHERE template = "'.$template.'" AND id_lang = "'.$this->context->language->id.'"');
```
Comment (Quetzacoalt91) [inline controllers/front/submitplus.php]: Please sanitize the values of this request, with an `(int)` for interger values and `pSQL(...)` for strings

**Comment by Quetzacoalt91** [comment]
> paypal.php, l. 1361:

``` php
$result = Db::getInstance()->ExecuteS('SELECT * FROM ' . _DB_PREFIX_ . 'paypal_order WHERE id_transaction = "' . $id_transaction . '"');
```

Add an `pSQL()` call with `$id_transaction`

File: paypal_plus/submit.php [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+					);
+					$return['error'][] = 'Une erreur est survenue pendant le paiement.';
+				}
+			} elseif ($submit == 'confirmCancel') {
+
+				$paypal->validateOrder(
+					$id_cart, getOrderStatus('order_canceled'),
+					$payment->transactions[0]->amount->total, $payment->payer->payment_method, null, $transaction
+				);
+				$return['success'][] = 'Votre commande a bien été annulé.';
```
Comment (Quetzacoalt91) [inline paypal_plus/submit.php]: Text should be translated thanks to your variable `$paypal`

File: paypal_plus/submit.php [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+			} else
+				$return['error'][] = 'Une erreur est survenue pendant le paiement.';
```
Comment (Quetzacoalt91) [inline paypal_plus/submit.php]: Text should be translated thanks to your variable `$paypal`

File: paypal_plus/submit.php [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+		} else
+			$return['error'][] = 'Une erreur est survenue pendant le paiement.';
```
Comment (Quetzacoalt91) [inline paypal_plus/submit.php]: Text should be translated thanks to your variable `$paypal`

File: paypal_plus/submit.php [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+	} else
+		$return['error'][] = 'Une erreur est survenue pendant le paiement.';
```
Comment (Quetzacoalt91) [inline paypal_plus/submit.php]: Text should be translated thanks to your variable `$paypal`

File: paypal_plus/submit.php [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+	function getOrderStatus($template)
+    {
+		global $cookie;
+        /*
+         * payment
+         * payment_error
+         * order_canceled
+         * refund
+         */
+        return Db::getInstance()->getValue('SELECT id_order_state FROM '._DB_PREFIX_.'order_state_lang WHERE template = "'.$template.'" AND id_lang = "'.$cookie->id_lang.'"');
```
Comment (Quetzacoalt91) [inline paypal_plus/submit.php]: Your values need to be sanitized

**Comment by Quetzacoalt91** [comment]
You should take care of your versioning. This PR seems to introduce a new minor version of the module instead of a patch. :1234:

File: views/templates/front/order-confirmation-plus-mobile.tpl [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+                        }
+                    });
+
+                    return false;
+                });
+
+            });
+
+        </script>
+        <div style="margin-top:15px;">
```
Comment (Quetzacoalt91) [inline views/templates/front/order-confirmation-plus-mobile.tpl]: Avoid style inline. These rules should be written in a CSS file

File: views/templates/front/order-confirmation-plus-mobile.tpl [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+                    <img src="{$img_dir}icon/order.gif" alt="{l s='Follow my order' mod='paypal'}" class="icon" />
+                {else}
+                    <i class="icon-chevron-left"></i>
+                {/if}</a>
+            <a href="{$link->getPageLink('history.php', true)}" title="{l s='Back to orders' mod='paypal'}" data-ajax="false">{l s='Back to orders' mod='paypal'}</a>
+        {/if}
+    </div>
+
+{elseif $state == 'failed' || $state == 'expired'}
+    <p class="alert alert-warning paypal-error">{l s="Une erreur est survenue pendant votre commande" mod='paypal'}</p>
```
Comment (Quetzacoalt91) [inline views/templates/front/order-confirmation-plus-mobile.tpl]: Text must be written in English first

File: views/templates/front/order-confirmation-plus-mobile.tpl [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+{elseif $state == 'canceled'}
+    <p class="alert alert-warning paypal-error">{l s="Votre commande a été annulé" mod='paypal'}</p>
```
Comment (Quetzacoalt91) [inline views/templates/front/order-confirmation-plus-mobile.tpl]: Same comment about the text

File: views/templates/front/order-confirmation-plus.tpl [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+                    <img src="{$img_dir}icon/order.gif" alt="{l s='Follow my order' mod='paypal'}" class="icon" />
+                {else}
+                    <i class="icon-chevron-left"></i>
+                {/if}</a>
+            <a href="{$link->getPageLink('history.php', true)}" title="{l s='Back to orders' mod='paypal'}" data-ajax="false">{l s='Back to orders' mod='paypal'}</a>
+        {/if}
+    </div>
+
+{elseif $state == 'failed' || $state == 'expired'}
+    <p class="alert alert-warning paypal-error">{l s="Une erreur est survenue pendant votre commande" mod='paypal'}</p>
```
Comment (Quetzacoalt91) [inline views/templates/front/order-confirmation-plus.tpl]: Same comment about the text

File: views/templates/front/order-confirmation-plus.tpl [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+{elseif $state == 'canceled'}
+    <p class="alert alert-warning paypal-error">{l s="Votre commande a été annulé" mod='paypal'}</p>
```
Comment (Quetzacoalt91) [inline views/templates/front/order-confirmation-plus.tpl]: And also here about the text

File: views/templates/hook/paypal_cart_summary.tpl [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+        <p style="margin-top:20px;">
+            - {l s='The total amount of your order is' mod='paypal'}
+            <span id="amount" class="price"><strong>{$total|escape:'htmlall':'UTF-8'}</strong></span> {if $use_taxes == 1}{l s='(tax incl.)' mod='paypal'}{/if}
+        </p>
+        <p>
+            - {l s='We accept the following currency to be sent by PayPal:' mod='paypal'}&nbsp;<b>{$currency->name|escape:'htmlall':'UTF-8'}</b>
+        </p>
+
+{if $useStyle14}
```
Comment (Quetzacoalt91) [inline views/templates/hook/paypal_cart_summary.tpl:83]: You introduce a new file **views/css/paypal_1_4.css** in this PR. This part should be moved inside then.

File: views/templates/hook/paypal_cart_summary.tpl [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+                    </td>
+                    <td>
+                        {$product.quantity}
+                    </td>
+                </tr>
+            {/foreach}
+            </table>
+        </div>
+
+        <p style="margin-top:20px;">
```
Comment (Quetzacoalt91) [inline views/templates/hook/paypal_cart_summary.tpl]: By writing CSS rules like this, merchants won't be able to set their own design rules. They will be always overridden

**Comment by thoma202** [comment]
@Quetzacoalt91 
As requested, I've sanitized and created dedicated css files for the style. 

Is it ok for you ?

**Comment by thoma202** [comment]
Hi @Quetzacoalt91 
Do you have any news on this ?

**Comment by mmoustai** [comment]
Hello,

You have a high number of errors who need to be solved before accepting this pull request for your major version of Paypal. 

download 22

Please correct this following section in priority of your module. [Red one]

Thanks you in advance.

**Comment by thoma202** [comment]
Hi @mmoustai 
We did fixed a lot of errors (almost 5000). 

For the few errors left those are errors that I don't know how to fix. 
image
If context exists we use it, this is just a backup in case something weird would happen. 

image

I'm still not sure what to do with this error ? 

image

When I try to change something it is not taken into account. 

image

Same here. moreover I can't find the two spaces validator is talking about. 

image

Concerning the escape of variables, this is complicated for us to change as it would mean a whole new test phase on our end, and we already have a lot of delay on the release of this module. 
Can we skip this so that we'll do it for the next release ? 

Thanks

**Pull request merged**